src/org/exist/indexing/StreamListener.java

```diff
@@ -43,6 +43,13 @@
         STORE,
 
         /**
+         * Mode for replacing the nodes of a document with another document
+         * this is really a group mode, which is later followed by {@link #REMOVE_ALL_NODES}
+         * and then {@link #STORE}
+         */
+        REPLACE_DOCUMENT,
```

Comment:
Previously when replacing a document the index calls would be:

```
setMode(REMOVE_ALL_NODES)
startIndexDocument()
...
endIndexDocument()

setMode(STORE)
startIndexDocument()
...
endIndexDocument()
```

The index had no way of knowing if the above sequence of events was the action of a document being replaced, or if it was two separate actions i.e. removing a document and then later storing a new document.

I needed to be able to differentiate between the two to perform the most efficient differential updates to my index.

With this PR, the the index calls for replacing a document are now:

```
setMode(REPLACE_DOCUMENT)
startReplaceDocument()

setMode(REMOVE_ALL_NODES)
startIndexDocument()
...
endIndexDocument()

setMode(STORE)
startIndexDocument()
...
endIndexDocument()

endReplaceDocument()
```